### ISSUE TITLE
Users

### ISSUE THREAD

**Issue by VadimSmD** [VadimSmD/KR#1]
Пользователи.
Поля:
Идентификатор
Ник
Пароль
Статус

Методы:
Добавление (регистрация)
Удаление
Изменение полей
Получение данных

**Comment by dankokin** [comment]
Поля: предлагаю добавить имя, фамилию, дату создания

Получение данных – получения одного пользователя? Всех не вижу смысла

**Comment by VadimSmD** [comment]
> Поля: предлагаю добавить имя, фамилию, дату создания
> 
> Получение данных – получения одного пользователя? Всех не вижу смысла

Да, сейчас подредактирую. Касательно получения данных - да это для 1 пользователя

**Comment by dankokin** [comment]
и еще, совсем забыл – авторизация пользователя на портале

**Comment by dankokin** [comment]
Как и планировали, начнем работу с пользователей.

- [ ] Создать ветку _features/users_
- [ ] С помощью утилиты golang-migrate создать миграцию с таблицей пользователей (подробнее: https://github.com/golang-migrate/migrate/blob/master/database/postgres/TUTORIAL.md раздел Create migrations). Папка с миграциями – internal/migrations/postgres
- [ ] Реализовать функцию для миграции БД, пример запуска миграции из кода приложения есть тут (https://github.com/evrone/go-clean-template/blob/master/internal/app/migrate.go)

Пойдем реализовывать пользователей снизу вверх по абстракции, то есть с функций для работы с БД

- [ ] Описать на Go структуру пользователя в _internal/entity/user_
- [ ] Создать _internal/repo/user.go_, в нем объявить структуру:
```go
type UserRepo struct {
     pg *pgxpool.Pool
}
```

- [ ] Реализовать для нее конструктор ```func NewUserRepo( pg *pgxpool.Pool) UserRepo```
- [ ] Реализовать создание пользователя:
```go
func (r Repo)  Insert(ctx context.Context, user entity.User) (entity.User, error)
```
- [ ] Скинуть мне на код-ревью

**Comment by dankokin** [comment]
Продолжаем работу над репозиторием пользователей. На основе реализованного ```Insert``` добить оставшиеся методы.
Методы для реализации:
- [ ] Добавление (регистрация) – подправить замечания;
- [ ] Удаление – здесь ничего сложного, удаление по идентификатору пользователя;
- [ ] Изменение полей – посмотри про разницу HTTP-методов ```PUT``` и ```PATCH``` и реализуй либо редактирование всех полей, либо части;
- [ ] Получение данных – при авторизации мы получаем пользователя по паролю и логину, в случае просмотра профиля пользователя у нас есть только его идентификатор, репозиторный метод хочется один, чтобы не дублировать код, подумай как это можно реализовать.
- [ ] Авторизация на портале – см. выше

По завершению этапа
- [ ] Ревью

**Comment by dankokin** [comment]
Продолжаем работу с пользователем:

- [ ] Необходимо реализовать слой с бизнес-логикой (service/usecase, как тебе кайф). На данный момент – это буквально вызов интерфейса репозитория. 

Над ним располагается транспортный слой, в нашем случае это HTTP. Для его реализации: 

- [x] Опиши документацию к сервису в формате [OpenAPI](https://spec.openapis.org/oas/latest.html);
- [x] Сгенерируй документ с помощью [Readocly](https://redocly.com/docs/cli). Необязательно скачивать npm и прочие непотребства (ведь ты белый человек, backend-разработчик) можно воспользоваться docker-образом
Команды будут выглядеть примерно так:
Генерация кода
`docker run --rm -v ${PWD}/docs:/spec redocly/cli build-docs -o docs/openapi.html openapi.yaml`
Линтер для документа
`docker run --rm -v ${PWD}/docs:/spec redocly/cli lint openapi.yaml`
Подробнее [тут](https://redocly.com/docs/cli/installation)

- [x] После того, как документ составлен, время генерировать код для транспортного уровня. Делаем это с [ogen](https://github.com/ogen-go/ogen);
- [x] Как только код реализован, самое время реализовать первый Handler – пользовательский;
- [x] Собрать все вместе и запустить http-сервер;
- [ ] Описать docker-compose с приложением и PostgreSQL и запустить;
- [ ] Установить (если еще нет) Postman, импортировать туда коллекцию с запросами из сгенерированного html документа либо из OpenAPI yaml-файла.
- [ ] Отправить запросы на свой http-сервер

По завершению:

- [ ] Код-ревью
